man/lib/c/stat.man
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)stat.2	6.7 (Berkeley) 10/25/87
.\"
.TH STAT 2 "October 25, 1987"
.UC 4
.SH NAME
stat, lstat, fstat \- get file status
.SH SYNOPSIS
.nf
.ft B
#include <sys/types.h>
#include <sys/stat.h>
.PP
.ft B
stat(path, buf)
char *path;
struct stat *buf;
.PP
.ft B
lstat(path, buf)
char *path;
struct stat *buf;
.PP
.ft B
fstat(fd, buf)
int fd;
struct stat *buf;
.fi
.ft R
.SH DESCRIPTION
.I Stat
obtains information about the file
.IR path .
Read, write or execute
permission of the named file is not required, but all directories
listed in the path name leading to the file must be reachable.
.PP
.I Lstat
is like \fIstat\fP except in the case where the named file is a symbolic link,
in which case
.I lstat
returns information about the link,
while
.I stat
returns information about the file the link references.
.PP
.I Fstat
obtains the same information about an open file
referenced by the argument descriptor, such as would
be obtained by an \fIopen\fP call.
.PP
.I Buf
is a pointer to a
.I stat
structure into which information is placed concerning the file.
The contents of the structure pointed to by
.I buf
.PP
.nf
.ta 1i 1.7i 3.0i
     struct stat {
	dev_t	st_dev;	/* device inode resides on */
	ino_t	st_ino;	/* this inode's number */
	u_short	st_mode;	/* protection */
	short	st_nlink;	/* number or hard links to the file */
	uid_t	st_uid;	/* user-id of owner */
	gid_t	st_gid;	/* group-id of owner */
	dev_t	st_rdev;	/* the device type, for inode that is device */
	off_t	st_size;	/* total size of file */
	time_t	st_atime;	/* file last access time */
	int	st_spare1;
	time_t	st_mtime;	/* file last modify time */
	int	st_spare2;
	time_t	st_ctime;	/* file last status change time */
	int	st_spare3;
	long	st_blksize;	/* optimal blocksize for file system i/o ops */
	long	st_blocks;	/* actual number of blocks allocated */
	long	st_serverID;	/* Sprite ID of server holding inode */
	long	st_version;	/* Version number for file */
	long	st_userType;	/* User-settable attribute word */
	long	st_devServerID;	/* For devices, Sprite ID of device's machine */
    };
.fi
.DT
.PP
.TP 12
st_atime
Time when file data was last accessed.  Changed by the following system
calls:
.IR mknod (2),
.IR utimes (2),
and
.IR read (2).
For directories, the access time is changed each time a file name
lookup passes through the directory.
.TP 12
st_mtime
Time when data was last modified.
It is not set by changes of owner, group, link count, or mode.
Changed by the following system calls:
.IR mknod (2),
.IR utimes (2),
.IR write (2).
.TP 12
st_ctime
Time when file status was last changed.
It is set both both by writing and changing the i-node.
Changed by the following system calls:
.IR chmod (2)
.IR chown (2),
.IR link (2),
.IR mknod (2),
.IR rename (2),
.IR unlink (2),
.IR utimes (2),
.IR write (2).
.TP 12
st_version
Version number for the file.  It is incremented each time the file
is opened for writing.
.PP
The status information word \fIst_mode\fP has bits:
.nf
.in +5n
.ta 1.6i 2.5i 3i
#define S_IFMT	0170000	/* type of file */
#define\ \ \ \ S_IFDIR	0040000	/* directory */
#define\ \ \ \ S_IFCHR	0020000	/* character special */
#define\ \ \ \ S_IFBLK	0060000	/* block special */
#define\ \ \ \ S_IFREG	0100000	/* regular */
#define\ \ \ \ S_IFLNK	0120000	/* symbolic link */
#define\ \ \ \ S_IFSOCK	0140000	/* socket */
#define\ \ \ \ S_IFPDEV	0150000	/* pseudo-device */
#define\ \ \ \ S_IFRLNK	0160000	/* remote link */
#define S_ISUID	0004000	/* set user id on execution */
#define S_ISGID	0002000	/* set group id on execution */
#define S_IREAD	0000400	/* read permission, owner */
#define S_IWRITE	0000200	/* write permission, owner */
#define S_IEXEC	0000100	/* execute/search permission, owner */
.fi
.in -5n
.PP
The mode bits 0000070 and 0000007 encode group and
others permissions (see
.IR chmod (2)).
.SH "RETURN VALUE
Upon successful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
.I errno
is set to indicate the error.
.SH "ERRORS
.I Stat
and
.I lstat
will fail if one or more of the following are true:
.TP 15
[ENOTDIR]
A component of the path prefix is not a directory.
.TP 15
[EINVAL]
The pathname contains a character with the high-order bit set.
.TP 15
[ENAMETOOLONG]
A component of a pathname exceeded 255 characters,
or an entire path name exceeded 1023 characters.
.TP 15
[ENOENT]
The named file does not exist.
.TP 15
[EACCES]
Search permission is denied for a component of the path prefix.
.TP 15
[ELOOP]
Too many symbolic links were encountered in translating the pathname.
.TP 15
[EFAULT]
.I Buf
or
.I name
points to an invalid address.
.TP 15
[EIO]
An I/O error occurred while reading from or writing to the file system.
.PP
.I Fstat
will fail if one or both of the following are true:
.TP 15
[EBADF]
.I Fildes
is not a valid open file descriptor.
.TP 15
[EFAULT]
.I Buf
points to an invalid address.
.TP 15
[EIO]
An I/O error occurred while reading from or writing to the file system.
.SH CAVEAT
The fields in the stat structure currently marked 
.IR st_spare1 ,
.IR st_spare2 ,
and
.I st_spare3
are present in preparation for inode time stamps expanding
to 64 bits.  This, however, can break certain programs that
depend on the time stamps being contiguous (in calls to
.IR utimes (2)).
.SH "SEE ALSO"
chmod(2), chown(2), utimes(2)
.SH BUGS
Applying
.I fstat
to a socket (and thus to a pipe)
returns a zero'd buffer,
except for the blocksize field,
and a unique device and inode number.


src/daemons/ipServer/stat.c
/* 
 * stat.c --
 *	
 *	Routines to print statistics about the IP server.
 *
 * Copyright 1987 Regents of the University of California
 * All rights reserved.
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Header: /sprite/src/daemons/ipServer/RCS/stat.c,v 1.5 89/04/10 16:36:03 mgbaker Exp $ SPRITE (Berkeley)";
#endif not lint


#include "sprite.h"
#include "ipServer.h"
#include "stat.h"

#include "time.h"
#include "proc.h"
#include "dev/net.h"

Stat_Info	stats;

static Proc_ResUsage	lastUsage;
static void ComputeDiff();
static char statDumpFile[100];


/*
 *----------------------------------------------------------------------
 *
 * Stat_Command --
 *
 *	Interprets IOC_NET_STATS command.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Statistics may be updated or printed.
 *
 *----------------------------------------------------------------------
 */

extern char myHostName[];

int
Stat_Command(command)
    unsigned int command;
{

    switch (command & 0xFFFF) { 
	case NET_STATS_RESET:
	    bzero((Address) &stats,sizeof(stats));
#ifndef KERNEL
	    Sys_GetTimeOfDay(&stats.startTime, (int *) NULL, (Boolean *) NULL);
	    Proc_GetResUsage(PROC_MY_PID, &lastUsage);
#endif
	    break;

	case NET_STATS_DUMP:
#ifndef KERNEL
	    (void) sprintf(statDumpFile, 
			"/tmp/ipServer.%s.%d", myHostName, command >> 16);
#endif
	    Stat_PrintInfo(-1, -1);
	    break;
    }
}


/*
 *----------------------------------------------------------------------
 *
 * Stat_PrintInfo --
 *
 *	Prints a summary of the statistics collected so far.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Statistics are printed on the standard error stream or to a file.
 *
 *----------------------------------------------------------------------
 */

/*ARGSUSED*/
int
Stat_PrintInfo(sigNum, sigCode)
    int		sigNum;		/* > 0 if called by signal handler. */
    int		sigCode;	/* used as stat file version #. */
{
    Proc_ResUsage	usage;
    Proc_ResUsage	diff;
    Time		curTime;
    int			localOffset;
    char		timeString[TIME_CVT_BUF_SIZE];
#ifndef KERNEL
    extern unsigned int fsNumTimeoutEvents;
    extern unsigned int fsNumStreamEvents;
    FILE		*stream;
    if (sigNum == -1) {
	stream = fopen(statDumpFile, "w");
    } else {
	stream = stderr;
    }
#else
    int stream = 0;
#endif

    Sys_GetTimeOfDay(&curTime, &localOffset, (Boolean *) NULL);
#ifndef sun4
    {
	char	*Version();
	(void) fprintf(stream, "IP Server stats:  %s\n\n", (char *) Version());
    }
#endif sun4

    Time_ToAscii(stats.startTime.seconds + 60 * (localOffset+60), 
		FALSE, timeString);
    (void) fprintf(stream, "Started at:   %s\n", timeString);

    Time_ToAscii(curTime.seconds + 60 * (localOffset+60), FALSE, timeString);
    (void) fprintf(stream, "Current time: %s\n", timeString);

    Time_Subtract(curTime, stats.startTime, &curTime);
    Time_ToAscii(curTime.seconds, TRUE, timeString);
    (void) fprintf(stream, "Difference:   %s\n", timeString);

#ifndef KERNEL
    Proc_GetResUsage(PROC_MY_PID, &usage);
    ComputeDiff(&usage, &lastUsage, &diff);
    (void) fprintf(stream, "\nResource Usage:\n");
	(void) fprintf(stream, " %16s %16s %11s %11s\n",
	    "kernel CPU", "user CPU", "invol.cs", "vol.cs");
	(void) fprintf(stream, " %9d.%06d %9d.%06d %11d %11d\n",
	    diff.kernelCpuUsage.seconds,
	    diff.kernelCpuUsage.microseconds,
	    diff.userCpuUsage.seconds,
	    diff.userCpuUsage.microseconds,
	    diff.numQuantumEnds,
	    diff.numWaitEvents);
#endif
    (void) fprintf(stream, "Socket:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "open","close", "read", "write", "ioctl", "select");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.sock.open,
	    stats.sock.close,
	    stats.sock.read,
	    stats.sock.write,
	    stats.sock.ioctl,
	    stats.sock.select);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "append", "app.part", "part.bytes", "app.fail", "remove", "fetch");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.sock.buffer.append,
	    stats.sock.buffer.appendPartial,
	    stats.sock.buffer.appPartBytes,
	    stats.sock.buffer.appendFail,
	    stats.sock.buffer.remove,
	    stats.sock.buffer.fetch);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "copies", "bytes copied", "rte $hits %", "dispatches", 
	    "timeouts", "streams");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.sock.buffer.copy,
	    stats.sock.buffer.copyBytes,
	    (stats.misc.routeCalls != 0 ? 
		(100*stats.misc.routeCacheHits)/stats.misc.routeCalls : 0),
	    stats.misc.dispatchLoop,
#ifndef KERNEL
	    fsNumTimeoutEvents,
	    fsNumStreamEvents);
#else
	    0, 0);
#endif


    (void) fprintf(stream, "\n");

    (void) fprintf(stream, "TCP:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "send: total", "data bytes", "data pack","ack only", "win probe", 
	    "win update");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.tcp.send.total,
	    stats.tcp.send.byte,
	    stats.tcp.send.pack,
	    stats.tcp.send.acks,
	    stats.tcp.send.probe,
	    stats.tcp.send.winUpdate);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "recv: total","bad", "data: ok","dupl.", "part dupl.", "bad order");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.tcp.recv.total,
	    stats.tcp.recv.badChecksum +
		stats.tcp.recv.badOffset +
		stats.tcp.recv.shortLen,
	    stats.tcp.recv.pack,
	    stats.tcp.recv.dupPack,
	    stats.tcp.recv.partDupPack,
	    stats.tcp.recv.ooPack);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "data bytes", "data > win", "after close", "win probe", 
	    "win update","");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.tcp.recv.byte,
	    stats.tcp.recv.packAfterWin,
	    stats.tcp.recv.afterClose,
	    stats.tcp.recv.winProbe,
	    stats.tcp.recv.winUpd,
	    0);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "acks", "dupl. ack", "ack > win", "urgent", "urg only","");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.tcp.recv.ackPack,
	    stats.tcp.recv.dupAck,
	    stats.tcp.recv.ackTooMuch,
	    stats.tcp.recv.urgent,
	    stats.tcp.recv.urgentOnly,
	    0);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "timer: tot.", "delay ack","retrans", "retr.drop", "persist", 
	    "2*msl");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.tcp.timerCalls,
	    stats.tcp.delayAck,
	    stats.tcp.rexmtTimeout,
	    stats.tcp.timeoutDrop,
	    stats.tcp.persistTimeout,
	    stats.tcp.mslTimeout);
	(void) fprintf(stream, " %11s %11s %11s\n",
	    "total keeps", "dropped", "probes");
	(void) fprintf(stream, " %11d %11d %11d\n",
	    stats.tcp.keepTimeout,
	    stats.tcp.keepDrops,
	    stats.tcp.keepProbe);

    (void) fprintf(stream, "\n");
    (void) fprintf(stream, "UDP:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "recv: total", "accepts", "deamons", "bad", "#bytes", "acc.bytes");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.udp.recv.total,
	    stats.udp.recv.accepted,
	    stats.udp.recv.daemon,
	    stats.udp.recv.shortLen+
		stats.udp.recv.badChecksum,
	    stats.udp.recv.dataLen,
	    stats.udp.recv.acceptLen);
	(void) fprintf(stream, " %11s %11s\n",
	    "send: total", "#bytes");
	(void) fprintf(stream, " %11d %11d\n",
	    stats.udp.send.total,
	    stats.udp.send.dataLen);

    (void) fprintf(stream, "\n");
    (void) fprintf(stream, "ICMP:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s %11s\n",
	    "recv: total", "short", "bad sum", "bad type", "bad code");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d\n",
	    stats.icmp.total, 
	    stats.icmp.shortLen, 
	    stats.icmp.badChecksum, 
	    stats.icmp.badType, 
	    stats.icmp.badCode);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s\n",
	  "unreachable", "redirect", "src quench", "time exceed", "param prob");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d\n",
	    stats.icmp.inHistogram[NET_ICMP_UNREACHABLE],
	    stats.icmp.inHistogram[NET_ICMP_REDIRECT],
	    stats.icmp.inHistogram[NET_ICMP_SOURCE_QUENCH],
	    stats.icmp.inHistogram[NET_ICMP_TIME_EXCEED],
	    stats.icmp.inHistogram[NET_ICMP_PARAM_PROB]);
	(void) fprintf(stream, " %11s %11s %11s %11s\n",
	    "echo", "timestamp", "info", "mask");
	(void) fprintf(stream, " %11d %11d %11d %11d\n",
	    stats.icmp.inHistogram[NET_ICMP_ECHO],
	    stats.icmp.inHistogram[NET_ICMP_TIMESTAMP],
	    stats.icmp.inHistogram[NET_ICMP_INFO_REQ],
	    stats.icmp.inHistogram[NET_ICMP_MASK_REQ]);

    (void) fprintf(stream, "\n");
    (void) fprintf(stream, "IP:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s %11s\n",
	    "recv: total", "frags","short", "bad sum", "not for us");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d\n",
	    stats.ip.totalRcv,
	    stats.ip.fragsRcv,
	    stats.ip.shortPacket + stats.ip.shortHeader + 
		stats.ip.shortLen,
	    stats.ip.badChecksum,
	    stats.ip.forwards);
	(void) fprintf(stream, " %11s %11s %11s %11s %11s %11s\n",
	    "send: total", "whole","fragmented", "#frags", "dont frag",
	    "TO calls");
	(void) fprintf(stream, " %11d %11d %11d %11d %11d %11d\n",
	    stats.ip.wholeSent+ stats.ip.fragOnSend,
	    stats.ip.wholeSent,
	    stats.ip.fragOnSend,
	    stats.ip.fragsSent,
	    stats.ip.dontFragment,
	    stats.ip.fragTimeouts);

    (void) fprintf(stream, "\n");
    (void) fprintf(stream, "Raw:\n");
	(void) fprintf(stream, " %11s %11s %11s %11s\n",
	    "recv: total", "accepted","send: total", "#bytes");
	(void) fprintf(stream, " %11d %11d %11d %11d\n",
	    stats.raw.recv.total,
	    stats.raw.recv.accepted,
	    stats.raw.send.total,
	    stats.raw.send.dataLen);
    (void) fflush(stream);
}

static void
ComputeDiff(newPtr, oldPtr, diffPtr)
    register Proc_ResUsage	*newPtr;
    register Proc_ResUsage	*oldPtr;
    register Proc_ResUsage	*diffPtr;
{
    Time_Subtract(newPtr->kernelCpuUsage, oldPtr->kernelCpuUsage, 
			&diffPtr->kernelCpuUsage);
    Time_Subtract(newPtr->userCpuUsage, oldPtr->userCpuUsage, 
			&diffPtr->userCpuUsage);
    diffPtr->numQuantumEnds = newPtr->numQuantumEnds - oldPtr->numQuantumEnds;
    diffPtr->numWaitEvents  = newPtr->numWaitEvents  - oldPtr->numWaitEvents;
}
